File: sdks/python/src/amlexia/cost.py
"""Token cost estimation (mirrors @amlexiahq/shared pricing table)."""

from __future__ import annotations

from typing import Any, Dict, Optional, Tuple

MODEL_PRICES: Dict[str, Dict[str, float]] = {
    "gpt-4o-mini": {"input": 0.15, "output": 0.6},
    "gpt-4o": {"input": 2.5, "output": 10},
    "claude-3-5-haiku": {"input": 0.8, "output": 4},
    "claude-3-5-sonnet": {"input": 3, "output": 15},
    "gemini-2.0-flash": {"input": 0.1, "output": 0.4},
}

PROVIDER_DEFAULT = {
    "openai": "gpt-4o-mini",
    "anthropic": "claude-3-5-haiku",
    "google": "gemini-2.0-flash",
}


def estimate_cost_usd(
    *,
    cost_usd: Optional[float] = None,
    model_name: Optional[str] = None,
    provider: Optional[str] = None,
    tokens_input: Optional[int] = None,
    tokens_output: Optional[int] = None,
    total_tokens: Optional[int] = None,
) -> Tuple[Optional[float], Optional[str]]:
    if cost_usd is not None and cost_usd > 0:
        return cost_usd, "reported"

    key = (model_name or "").lower()
    price = MODEL_PRICES.get(key)
    if not price and provider:
        default = PROVIDER_DEFAULT.get(provider.lower())
        if default:
            price = MODEL_PRICES.get(default)
    if not price:
        return cost_usd, None

    inp = tokens_input or 0
    out = tokens_output or 0
    if inp == 0 and out == 0:
        if not total_tokens:
            return cost_usd, None
        blended = (price["input"] + price["output"]) / 2
        return (total_tokens / 1_000_000) * blended, "estimated"

    est = (inp / 1_000_000) * price["input"] + (out / 1_000_000) * price["output"]
    return round(est, 8), "estimated"


def enrich_event_cost(event: Dict[str, Any]) -> Dict[str, Any]:
    cost, _source = estimate_cost_usd(
        cost_usd=event.get("cost_usd"),
        model_name=event.get("model_name"),
        provider=event.get("provider") or event.get("provider_name"),
        tokens_input=event.get("tokens_input"),
        tokens_output=event.get("tokens_output"),
        total_tokens=event.get("total_tokens"),
    )
    if cost is not None:
        event["cost_usd"] = cost
    return event

File: sdks/ruby/lib/amlexia/cost.rb
# frozen_string_literal: true

module Amlexia
  MODEL_PRICES = {
    "gpt-4o-mini" => { input: 0.15, output: 0.6 },
    "gpt-4o" => { input: 2.5, output: 10 },
    "claude-3-5-haiku" => { input: 0.8, output: 4 },
    "claude-3-5-sonnet" => { input: 3, output: 15 },
    "gemini-2.0-flash" => { input: 0.1, output: 0.4 }
  }.freeze

  PROVIDER_DEFAULT_MODEL = {
    "openai" => "gpt-4o-mini",
    "anthropic" => "claude-3-5-haiku",
    "google" => "gemini-2.0-flash"
  }.freeze

  module Cost
    module_function

    def estimate_cost_usd(cost_usd:, model_name: nil, provider: nil, tokens_input: nil, tokens_output: nil, total_tokens: nil)
      return [cost_usd, "reported"] if cost_usd && cost_usd.positive?

      key = (model_name || "").downcase
      price = MODEL_PRICES[key]
      price ||= MODEL_PRICES[PROVIDER_DEFAULT_MODEL[(provider || "").downcase]] if provider
      return [cost_usd, nil] unless price

      inp = tokens_input || 0
      out = tokens_output || 0
      if inp.zero? && out.zero?
        return [cost_usd, nil] unless total_tokens&.positive?

        blended = (price[:input] + price[:output]) / 2.0
        return [(total_tokens / 1_000_000.0) * blended, "estimated"]
      end

      est = (inp / 1_000_000.0) * price[:input] + (out / 1_000_000.0) * price[:output]
      [est.round(8), "estimated"]
    end

    def enrich_event(event)
      cost, = estimate_cost_usd(
        cost_usd: event[:cost_usd],
        model_name: event[:model_name],
        provider: event[:provider] || event[:provider_name],
        tokens_input: event[:tokens_input],
        tokens_output: event[:tokens_output],
        total_tokens: event[:total_tokens]
      )
      event[:cost_usd] = cost if cost
      event
    end
  end
end
